modules/afprog/afprog.c

```diff
@@ -358,18 +430,55 @@ afprogram_dd_init(LogPipe *s)
     }
   log_pipe_append(&self->super.super.super, (LogPipe *) self->writer);
 
-  return afprogram_dd_reopen(self);
+  return restore_successful ? TRUE : afprogram_dd_reopen(self);
+}
+
+static inline void
+afprogram_dd_store_reload_info(AFProgramDestDriver * const self, GlobalConfig * const cfg)
+{
```

Comment:
See the following example:

void f(int * const i) {
 *i += 42;
}

If you made a typo anywhere in the body to forget the star, const would
help the compiler to catch the issue.


On Thu, Oct 8, 2015 at 10:38 AM, Nagy, Tamás <tamas.nagy@balabit.com> wrote:

> Recall from yesterday's meeting that security and program correctness is a
> growing concern. I welcome all efforts that could make it easier to write
> bug free code.
>
> It all depends on the team's preferences. If others say that it doesn't
> add enough value while at the same time makes the code harder to read, I
> must bow before their will.
>
> Here's my take in this exact question. For private functions it should be
> allowed.
>
> They say that for an API, specifying it at the prototype leaks
> implementation details. Do note that constity of local primitive types is
> not part of the contract.
>
> /* Examine the following valid code */
>
> int f(int * i);
> int f(int * const i) {
>  return *i + 42;
> }
>
> /* While this one fails to compile */
>
> int f(int * i);
> int f(const int * i) {
>  return *i + 42;
> }
>
> So we could have our cake and eat it by keeping the API the same, while
> providing the little extra information at the definition that Ádám
> mentioned earlier. The compiler still catches modification attempts to the
> read only arguments this way.
>
> I've proposed this to Ádám, but he kind of dislikes the idea of using
> mismatching prototype/implementation. What do the others think?
>
>
> On Wed, Oct 7, 2015 at 5:40 PM, MÓZES Ádám István <
> notifications@github.com> wrote:
>
>> In modules/afprog/afprog.c
>> <https://github.com/balabit/syslog-ng/pull/722#discussion_r41406196>:
>>
>> > @@ -358,18 +430,55 @@ afprogram_dd_init(LogPipe *s)
>> >      }
>> >    log_pipe_append(&self->super.super.super, (LogPipe *) self->writer);
>> >
>> > -  return afprogram_dd_reopen(self);
>> > +  return restore_successful ? TRUE : afprogram_dd_reopen(self);
>> > +}
>> > +
>> > +static inline void
>> > +afprogram_dd_store_reload_info(AFProgramDestDriver * const self, GlobalConfig * const cfg)
>> > +{
>>
>> Since the compilers can prove the constness, this notation has effects
>> only on humans.
>> For example, when I see this constant pointer, I won't expect to change
>> its value, therefore I do not need to follow the changes of this variable.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/balabit/syslog-ng/pull/722/files#r41406196>.
>>
>
>